doc/contribute/coding_guidelines/index.rst

```diff
@@ -1297,6 +1297,188 @@ macros. Given that Zephyr uses a fork of the corresponding upstream for each
+   memmove(),ISO/IEC 9899:2011
+   memset(),ISO/IEC 9899:2011
+   perror(),ISO/IEC 9899:2011
+   printf(),ISO/IEC 9899:2011
+   putc(),ISO/IEC 9899:2011
+   putchar(),ISO/IEC 9899:2011
+   puts(),ISO/IEC 9899:2011
+   qsort(),ISO/IEC 9899:2011
+   `qsort_r()`_,GNU
+   rand(),ISO/IEC 9899:2011
+   realloc(),ISO/IEC 9899:2011
+   `reallocarray()`_,OpenBSD
+   snprintf(),ISO/IEC 9899:2011
+   sprintf(),ISO/IEC 9899:2011
+   sqrt(),ISO/IEC 9899:2011
+   sqrtf(),ISO/IEC 9899:2011
+   srand(),ISO/IEC 9899:2011
+   strcat(),ISO/IEC 9899:2011
+   strchr(),ISO/IEC 9899:2011
+   strcmp(),ISO/IEC 9899:2011
+   strcpy(),ISO/IEC 9899:2011
+   strcspn(),ISO/IEC 9899:2011
+   strerror(),ISO/IEC 9899:2011
+   `strerror_r()`_,POSIX.1-2001
+   strlen(),ISO/IEC 9899:2011
+   strncat(),ISO/IEC 9899:2011
+   strncmp(),ISO/IEC 9899:2011
+   strncpy(),ISO/IEC 9899:2011
+   `strnlen()`_,POSIX.1-2008
+   strrchr(),ISO/IEC 9899:2011
+   strspn(),ISO/IEC 9899:2011
+   strstr(),ISO/IEC 9899:2011
+   `strtok_r()`_,POSIX.1-2001
+   strtol(),ISO/IEC 9899:2011
+   strtoll(),ISO/IEC 9899:2011
+   strtoul(),ISO/IEC 9899:2011
+   strtoull(),ISO/IEC 9899:2011
+   time(),ISO/IEC 9899:2011
+   tolower(),ISO/IEC 9899:2011
+   toupper(),ISO/IEC 9899:2011
+   vfprintf(),ISO/IEC 9899:2011
+   vprintf(),ISO/IEC 9899:2011
+   vsnprintf(),ISO/IEC 9899:2011
+   vsprintf(),ISO/IEC 9899:2011
+
+All of the functions listed above must be implemented by the
+:ref:`minimal libc <c_library_minimal>` to ensure that the Zephyr kernel can
+build with the minimal libc.
+
+In addition, any functions from the above list that are not part of the
+ISO/IEC 9899:2011 standard must be implemented by the
+:ref:`common libc <c_library_common>` to ensure their availability across
+multiple C standard libraries.
+
+The "Zephyr kernel" in this context consists of the following components:
+
+* Kernel (:file:`kernel`)
+* OS Library (:file:`lib/os`)
+* Architecture Port (:file:`arch`)
+* Logging Subsystem (:file:`subsys/logging`)
+
+Rationale
+---------
+
+Zephyr kernel must be able to build with the
+:ref:`minimal libc <c_library_minimal>`, a limited C standard library
+implementation that is part of the Zephyr RTOS and maintained by the Zephyr
+Project, to allow self-contained testing and verification of the kernel and
+core OS services.
+
+In order to ensure that the Zephyr kernel can build with the minimal libc, it
+is necessary to restrict the use of the C standard library functions and macros
+in the Zephyr kernel to the functions and macros that are available as part of
+the minimal libc.
+
+Rule A.5: C Standard Library Usage Restrictions in Zephyr Codebase
+==================================================================
+
+Severity
+--------
+
+Required
+
+Description
+-----------
+
+The use of the C standard library functions and macros in the Zephyr codebase
+shall be limited to the functions from the ISO/IEC 9899:2011 standard, also
+known as C11, unless exempted by this rule.
```

Comment:
@keith-packard Speaking of the _optional_ features in https://github.com/zephyrproject-rtos/zephyr/pull/57598#issuecomment-1550304383, I should probably revise the Rule A.5 to refer to the _mandatory_  functions and macros.

```suggestion
The use of the C standard library functions and macros in the Zephyr codebase
shall be limited to the mandatory functions from the ISO/IEC 9899:2011 standard, also
known as C11, unless exempted by this rule.
```

Reply:
Yeah, we definitely don't want to stick annex K in here. We probably do want `<math.h>` though, and that's in annex F (given the initial, one suspects a bit of careful annex ordering here), and probably `<complex.h>`, which is in annex G. Maybe just exclude annex K?

Reply:
Done. https://github.com/zephyrproject-rtos/zephyr/pull/57598#issuecomment-1553424876